sql/log.cc

```diff
@@ -2192,77 +2154,168 @@ bool Slow_log_throttle::log(THD *thd, bool eligible)
+  DBUG_RETURN(max_found);
 }
 
-int File_query_log::new_file()
+bool File_query_log::set_rotated_name(const bool need_lock)
 {
-  int error= 0, close_on_error= FALSE;
-  char new_name[FN_REFLEN], *old_name;
+  DBUG_ENTER("File_query_log::set_rotated_name");
+  if (need_lock) mysql_mutex_assert_owner(&LOCK_log);
 
-  DBUG_ENTER("File_query_log::new_file");
-  if (!is_open())
+  if (!max_slowlog_size)
   {
-    DBUG_PRINT("info",("log is closed"));
-    DBUG_RETURN(error);
+    fn_format(log_file_name, name, mysql_data_home, "", MY_UNPACK_FILENAME);
+    DBUG_RETURN(false);
   }
 
-  mysql_mutex_assert_owner(&LOCK_log);
-
-  if ((cur_log_ext == (ulong)-1) || max_slowlog_size == 0)
+  if (cur_log_ext == 0)
   {
-    strcpy(new_name, name);
-    if ((error= generate_new_log_name(new_name, &cur_log_ext, name, false)))
-      goto end;
+    fn_format(log_file_name, name, mysql_data_home, "", MY_UNPACK_FILENAME);
+    cur_log_ext = get_last_extension(log_file_name) + 1;
   }
   else
   {
-    if (cur_log_ext == MAX_LOG_UNIQUE_FN_EXT)
+    cur_log_ext++;
+  }
+
+  if (cur_log_ext > 0)
+  {
+    /* check if reached the maximum possible extension number */
+    if (cur_log_ext >= MAX_LOG_UNIQUE_FN_EXT)
+    {
+      sql_print_error("Log filename extension number exhausted: %06lu. \
+                       Please fix this by archiving old logs and \
+                       updating the index files.", cur_log_ext);
+      DBUG_RETURN(true);
+    }
+
+    if (snprintf(log_file_name, sizeof(log_file_name), "%s.%06lu", name,
+                 cur_log_ext) >= static_cast<int>(sizeof(log_file_name)))
     {
-      error= 1;
-      goto end;
+      my_printf_error(ER_NO_UNIQUE_LOGFILE,
+                      ER_THD(current_thd, ER_NO_UNIQUE_LOGFILE),
+                      MYF(ME_FATALERROR), name);
+      sql_print_error(ER(ER_NO_UNIQUE_LOGFILE), name);
+      DBUG_RETURN(true);
     }
-    snprintf(new_name, sizeof(new_name), "%s.%06lu", name, ++cur_log_ext);
+
+    if (need_lock) mysql_mutex_lock(&LOCK_global_system_variables);
+    opt_slow_logname = my_strdup(key_memory_LOG_name, log_file_name, MYF(MY_WME));
+    if (need_lock) mysql_mutex_unlock(&LOCK_global_system_variables);
   }
 
-  /*
-    close will try to free name and zero name pointer,
-    We saving current name value and zeroing the pointer to
-    prvent it.
-  */
-  old_name= name;
-  name= NULL;
-  close();
-  name= old_name;
+  DBUG_RETURN(false);
+}
+
+bool File_query_log::rotate(const ulong max_size)
+{
+  DBUG_ENTER("File_query_log::rotate");
+  mysql_mutex_assert_owner(&LOCK_log);
+
+  if (my_b_tell(&log_file) > max_size)
+  {
+    if (set_rotated_name(true)) DBUG_RETURN(true);
+    close();
+    if (open()) DBUG_RETURN(true);
+  }
+
+  DBUG_RETURN(false);
+}
+
+bool File_query_log::purge_logs()
+{
+  DBUG_ENTER("File_query_log::purge_logs");
+  mysql_mutex_assert_owner(&LOCK_log);
+
+  if (max_slowlog_files == 0 || cur_log_ext < 1 ||
+      last_removed_ext == cur_log_ext - 1)
```

Comment:
This `last_removed_ext == cur_log_ext - 1` is here to check that there is at least one more file except currently used one. Otherwise there is nothing to purge.
So if there were 3 files and you change max_slowlog_files to 1 this condition shouldn't prevent it from removing all except currently active file.